File: android/src/main/java/expo/modules/docusign/DocuSignFailure.kt
package expo.modules.docusign

import com.docusign.androidsdk.exceptions.DSException
import com.docusign.androidsdk.exceptions.DSRestException
import java.io.IOException
import org.json.JSONObject

/**
 * Facts about a runtime failure, sent to JS as flat keys.
 *
 * JS derives the failure's `reason` from these, so native code reports what it observed and never
 * guesses. Every field is optional because each failure knows a different subset.
 */
internal data class FailureDetails(
  val nativeDomain: String? = null,
  val nativeCode: String? = null,
  val nativeMessage: String? = null,
  val underlyingDomain: String? = null,
  val underlyingCode: String? = null,
  val underlyingMessage: String? = null,
  val httpStatus: Int? = null,
  val docusignErrorCode: String? = null,
  val docusignMessage: String? = null
) {
  /** Records a lower-level error without overwriting one the SDK already reported. */
  fun withUnderlyingIfAbsent(error: Throwable): FailureDetails {
    if (underlyingDomain != null) return this
    val root = rootCause(error) ?: error
    return copy(
      underlyingDomain = root.javaClass.name,
      underlyingCode = sdkErrorCode(root),
      underlyingMessage = root.message
    )
  }

  fun withHttp(error: DocuSignHttpException): FailureDetails =
    copy(
      httpStatus = error.status,
      docusignErrorCode = error.docusignErrorCode,
      docusignMessage = error.docusignMessage
    )

  fun toMap(): Map<String, Any> =
    listOfNotNull<Pair<String, Any>>(
      nativeDomain?.let { "nativeDomain" to it },
      nativeCode?.let { "nativeCode" to it },
      nativeMessage?.let { "nativeMessage" to it },
      underlyingDomain?.let { "underlyingDomain" to it },
      underlyingCode?.let { "underlyingCode" to it },
      underlyingMessage?.let { "underlyingMessage" to it },
      httpStatus?.let { "httpStatus" to it },
      docusignErrorCode?.let { "docusignErrorCode" to it },
      docusignMessage?.let { "docusignMessage" to it }
    ).toMap()

  companion object {
    /**
     * Reads the SDK's own error code and message where it provides them. The exception message
     * alone was all this module forwarded before, and it rarely says what failed.
     */
    fun from(error: Throwable): FailureDetails {
      val cause = rootCause(error)
      return FailureDetails(
        nativeDomain = error.javaClass.name,
        nativeCode = sdkErrorCode(error),
        nativeMessage = (error as? DSException)?.errorMsg?.takeIf { it.isNotBlank() } ?: error.message,
        underlyingDomain = cause?.javaClass?.name,
        underlyingCode = cause?.let { sdkErrorCode(it) },
        underlyingMessage = cause?.message,
        httpStatus = (error as? DSRestException)?.responseCode?.takeIf { it > 0 }
      )
    }

    private fun sdkErrorCode(error: Throwable): String? =
      (error as? DSException)?.errorCode?.takeIf { it.isNotBlank() }

    private const val MAX_CAUSE_DEPTH = 8

    /**
     * The deepest cause, bounded and safe against cycles. Java wraps transport failures, so the
     * SocketTimeoutException or UnknownHostException that explains a failure sits at the root of
     * the chain, often more than one level below the exception the SDK hands over. iOS keeps the
     * immediate underlying error instead, because the deepest error under an NSURLErrorDomain
     * failure is a CFNetwork error that the reason rules do not classify as a network failure.
     */
    private fun rootCause(error: Throwable): Throwable? {
      val seen = mutableSetOf(error)
      var current = error.cause ?: return null
      repeat(MAX_CAUSE_DEPTH) {
        seen.add(current)
        val next = current.cause
        if (next == null || next in seen) return current
        current = next
      }
      return current
    }
  }
}

/**
 * A non-2xx response to a request this package makes itself. DocuSign's error body names the real
 * problem, such as `UNKNOWN_ENVELOPE_RECIPIENT` for a recipient that does not match the envelope,
 * so it is kept rather than reduced to the status line.
 */
internal class DocuSignHttpException(
  val status: Int,
  val docusignErrorCode: String?,
  val docusignMessage: String?
) : IOException("DocuSign request failed with HTTP $status") {
  companion object {
    fun from(status: Int, body: String): DocuSignHttpException {
      val json = runCatching { JSONObject(body) }.getOrNull()
      return DocuSignHttpException(
        status = status,
        docusignErrorCode = json?.nonEmpty("errorCode") ?: json?.nonEmpty("error"),
        docusignMessage = json?.nonEmpty("message") ?: json?.nonEmpty("error_description")
      )
    }

    private fun JSONObject.nonEmpty(key: String): String? =
      if (isNull(key)) null else optString(key).takeIf { it.isNotBlank() }
  }
}

/**
 * A runtime failure, as opposed to a caller mistake.
 *
 * Caller mistakes are coded exceptions and reject. A rejection crosses the Expo bridge with only a
 * code and a message, so runtime failures resolve with this payload instead and JS turns it into a
 * thrown `DocuSignError` with every detail intact.
 */
internal class DocuSignFailure(
  val code: String,
  override val message: String,
  val details: FailureDetails,
  val envelopeId: String? = null
) : Exception(message) {
  fun toPayload(fallbackEnvelopeId: String?): Map<String, Any> {
    val payload = details.toMap().toMutableMap()
    payload["errorCode"] = code
    payload["errorMessage"] = message
    (envelopeId ?: fallbackEnvelopeId)?.takeIf { it.isNotEmpty() }?.let { payload["envelopeId"] = it }
    return payload
  }
}

File: ios/DocuSignFailure.swift
import Foundation

/// Facts about a runtime failure, sent to JS as flat keys.
///
/// JS derives the failure's `reason` from these, so native code reports what it observed and never
/// guesses. Every field is optional because each failure knows a different subset.
internal struct FailureDetails {
  var nativeDomain: String?
  var nativeCode: String?
  var nativeMessage: String?
  var underlyingDomain: String?
  var underlyingCode: String?
  var underlyingMessage: String?
  var httpStatus: Int?
  var docusignErrorCode: String?
  var docusignMessage: String?

  init() {}

  /// Reads the error's own domain and code plus `NSUnderlyingErrorKey`. The domain is kept because
  /// the SDK raises errors from several domains, so a code on its own is ambiguous.
  init(error: Error) {
    let nsError = error as NSError
    nativeDomain = nsError.domain
    nativeCode = String(nsError.code)
    nativeMessage = nsError.localizedDescription
    if let underlying = nsError.userInfo[NSUnderlyingErrorKey] as? NSError {
      setUnderlying(underlying)
    }
  }

  /// Records a lower-level error without overwriting one the SDK already reported.
  mutating func setUnderlyingIfAbsent(_ error: Error) {
    guard underlyingDomain == nil else { return }
    setUnderlying(error as NSError)
  }

  private mutating func setUnderlying(_ error: NSError) {
    underlyingDomain = error.domain
    underlyingCode = String(error.code)
    underlyingMessage = error.localizedDescription
  }

  /// Keeps DocuSign's own error code and message from a REST or OAuth error body, which is the
  /// only place a backend problem such as a mismatched recipient is named.
  mutating func setResponse(status: Int, body: Data?) {
    httpStatus = status
    guard
      let body = body,
      let json = try? JSONSerialization.jsonObject(with: body) as? [String: Any]
    else { return }
    docusignErrorCode = Self.nonEmpty(json["errorCode"]) ?? Self.nonEmpty(json["error"])
    docusignMessage = Self.nonEmpty(json["message"]) ?? Self.nonEmpty(json["error_description"])
  }

  private static func nonEmpty(_ value: Any?) -> String? {
    guard let text = value as? String, !text.isEmpty else { return nil }
    return text
  }

  var payload: [String: Any] {
    let values: [String: Any?] = [
      "nativeDomain": nativeDomain,
      "nativeCode": nativeCode,
      "nativeMessage": nativeMessage,
      "underlyingDomain": underlyingDomain,
      "underlyingCode": underlyingCode,
      "underlyingMessage": underlyingMessage,
      "httpStatus": httpStatus,
      "docusignErrorCode": docusignErrorCode,
      "docusignMessage": docusignMessage
    ]
    return values.compactMapValues { $0 }
  }
}

/// A runtime failure, as opposed to a caller mistake.
///
/// Caller mistakes are thrown as coded exceptions and reject. A rejection crosses the Expo bridge
/// with only a code and a message, so runtime failures resolve with this payload instead and JS
/// turns it into a thrown `DocuSignError` with every detail intact.
internal struct DocuSignFailure: Error {
  let code: String
  let message: String
  let details: FailureDetails
  let envelopeId: String?

  init(code: String, message: String, details: FailureDetails, envelopeId: String? = nil) {
    self.code = code
    self.message = message
    self.details = details
    self.envelopeId = envelopeId
  }

  func payload(fallbackEnvelopeId: String?) -> [String: Any] {
    var payload = details.payload
    payload["errorCode"] = code
    payload["errorMessage"] = message
    if let envelopeId = envelopeId ?? fallbackEnvelopeId, !envelopeId.isEmpty {
      payload["envelopeId"] = envelopeId
    }
    return payload
  }
}
